### PR TITLE
Add WKO retry improvement miscellaneous usescases

### PR DESCRIPTION
**Pull request by xiancao**
Add WKO retry improvement miscellaneous usescases

Jenkins result:
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/14319/ (The failed tests also failed in main branch)

### CODE REVIEW

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+    checkDomainStatusClusterConditionTypeHasExpectedStatus(domainUid, domainNamespace, clusterName,
+        DOMAIN_STATUS_CONDITION_AVAILABLE_TYPE, "False");
+
+    // wait for all the pods are up
+    // check admin server pod is ready
+    logger.info("Wait for admin server pod {0} to be ready in namespace {1}",
+        adminServerPodName, domainNamespace);
+    checkPodReadyAndServiceExists(adminServerPodName, domainUid, domainNamespace);
+
+    // check the domain status is Unavailable and Incomplete
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:138]: Even after the service and pod is available, why the completed status condition is false ? Is it due to reason the cluster service is not up ? If so please add the proper description

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+    domainNamespace = namespaces.get(1);
+
+    opServiceAccount = opNamespace + "-sa";
+    // install and verify operator with REST API
+    logger.info("Install an operator in namespace {0}, managing namespace {1}",
+        opNamespace, domainNamespace);
+    installAndVerifyOperator(opNamespace, opServiceAccount, true, 0, domainNamespace);
+  }
+
+  /**
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:98]: Can you explain the usecase in more detail, without copying the information from the wiki.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+
+    // check the cluster condition is Available and Complete
+    checkDomainStatusClusterConditionTypeHasExpectedStatus(shortRetryPolicy, domainUid, domainNamespace,
+        clusterName, DOMAIN_STATUS_CONDITION_COMPLETED_TYPE, "True", DOMAIN_VERSION);
+    checkDomainStatusClusterConditionTypeHasExpectedStatus(shortRetryPolicy, domainUid, domainNamespace,
+        clusterName, DOMAIN_STATUS_CONDITION_AVAILABLE_TYPE, "True", DOMAIN_VERSION);
+  }
+
+  /**
+   * Set cluster to NEVER. Expect domain status to stay available and complete.
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:179]: Again, describe the usecase in term of the code used, not the wiki page

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+    // check the cluster pod is deleted
+    for (int i = 1; i <= replicaCount; i++) {
+      logger.info("Wait for managed server pod {0} to be deleted in namespace {1}",
+          managedServerPrefix + i, domainNamespace);
+      checkPodDeleted(managedServerPrefix + i, domainUid, domainNamespace);
+    }
+  }
+
+  /**
+   * Set cluster back to IF_NEEDED, expect recovery, where cluster becomes Available once its one pod is ready
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:222]: What is recovery means here ?

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+        clusterName, DOMAIN_STATUS_CONDITION_AVAILABLE_TYPE, "True", DOMAIN_VERSION);
+    checkDomainStatusClusterConditionTypeHasExpectedStatus(withQuickRetryPolicy, domainUid, domainNamespace,
+        clusterName, DOMAIN_STATUS_CONDITION_COMPLETED_TYPE, "True", DOMAIN_VERSION);
+    checkDomainStatusConditionTypeHasExpectedStatus(domainUid, domainNamespace,
+        DOMAIN_STATUS_CONDITION_COMPLETED_TYPE, "True");
+    checkDomainStatusConditionTypeHasExpectedStatus(domainUid, domainNamespace,
+        DOMAIN_STATUS_CONDITION_AVAILABLE_TYPE, "True");
+  }
+
+  /**
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:302]: Please rewrite the javadoc with clear explanation

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+    logger.info("Updating cluster configuration using patch string: {0}", patchStr);
+    V1Patch patch = new V1Patch(patchStr);
+    assertTrue(patchClusterCustomResource(clusterName, domainNamespace,
+        patch, V1Patch.PATCH_FORMAT_JSON_PATCH), "Failed to patch cluster");
+
+    // check the domain Complete goes to false
+    checkDomainStatusConditionTypeHasExpectedStatus(shortRetryPolicy, domainUid, domainNamespace,
+        DOMAIN_STATUS_CONDITION_COMPLETED_TYPE, "False", DOMAIN_VERSION);
+
+    // Enable this check when OWLS-104622 is fixed
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:323]: No reference to OWLS in public code. You can update the JIRA how to verify the usecase after changing the code e.g add the following libe to tehi class to verify teh resolution 

checkDomainStatusClusterConditionTypeHasExpectedStatus(shortRetryPolicy, domainUid, domainNamespace,
   clusterName, DOMAIN_STATUS_CONDITION_COMPLETED_TYPE, "False", DOMAIN_VERSION)

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+    // verify Domain available should stay true throughout
+    checkDomainStatusConditionTypeHasExpectedStatus(withQuickRetryPolicy, domainUid, domainNamespace,
+        DOMAIN_STATUS_CONDITION_AVAILABLE_TYPE, "True", DOMAIN_VERSION);
+    // check the domain Complete back to True when the cluster shutdown completes
+    checkDomainStatusConditionTypeHasExpectedStatus(withQuickRetryPolicy, domainUid, domainNamespace,
+        DOMAIN_STATUS_CONDITION_COMPLETED_TYPE, "True", DOMAIN_VERSION);
+  }
+
+  /**
+   * Try set cluster replicas=2. The cluster Complete should immediately go false, and cluster Available should go true
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:355]: Rewrite the javadoc

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+   * the moment one of the cluster's members becomes ready, and cluster Complete should become true once both servers
+   * are ready. Domain available and complete should both flip to false, domain available should become true when
+   * cluster available becomes true, and domain complete should only come back to true when the cluster is fully up
+   * and ready.
+   */
+  @Order(6)
+  @Test
+  @DisplayName("Set cluster replicas to 2, verify the domain and cluster status Complete and Available condition")
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:363]: @DisplayName("Set cluster replicas to 2, verify the domain and cluster status Complete and Available condition") -->
@DisplayName("Verify the domain and cluster status Complete and Available condition after scaling up the cluster")

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 9]
```diff
@@ -0,0 +1,471 @@
...
+    // check cluster Complete becomes true once both servers are ready
+    checkDomainStatusClusterConditionTypeHasExpectedStatus(shortRetryPolicy, domainUid, domainNamespace,
+        clusterName, DOMAIN_STATUS_CONDITION_COMPLETED_TYPE, "True", DOMAIN_VERSION);
+    // check domain complete should only come back to true when the cluster is fully up
+    checkDomainStatusConditionTypeHasExpectedStatus(shortRetryPolicy, domainUid, domainNamespace,
+        DOMAIN_STATUS_CONDITION_COMPLETED_TYPE, "True", DOMAIN_VERSION);
+  }
+
+  /**
+   * Delete one of the cluster's pods. Cluster Available should stay true throughout, because the cluster remains
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:402]: Rewrite the javadoc

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+  }
+
+  /**
+   * Delete the Admin server pod. Expect domain status to flip to Complete false and Available false,
+   * but Cluster status to remain unchanged throughout, staying at Complete and Available.
+   * Once Admin server pod restarts, the Domain status should become true.
+   */
+  @Order(8)
+  @Test
+  @DisplayName("Delete the admin server pod. Verify the domain and cluster status")
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:444]: @DisplayName("Delete the admin server pod. Verify the domain and cluster status") --> DisplayName("Verify the domain and cluster status" in the absence of Admin Server)

**Review by anpanigr — COMMENTED** [review]
Modify javadoc on the test method.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+    // check cluster status condition Complete becomes true once both servers are ready
+    checkDomainStatusClusterConditionTypeHasExpectedStatus(shortRetryPolicy, domainUid, domainNamespace,
+        clusterName, DOMAIN_STATUS_CONDITION_COMPLETED_TYPE, "True", DOMAIN_VERSION);
+    // check domain status condition complete becomes true when the cluster is fully up
+    checkDomainStatusConditionTypeHasExpectedStatus(shortRetryPolicy, domainUid, domainNamespace,
+        DOMAIN_STATUS_CONDITION_COMPLETED_TYPE, "True", DOMAIN_VERSION);
+  }
+
+  /**
+   * Test domain and cluster status conditions when deleting one of the cluster server pods.
```
[the remark this answers is not in the document]
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:430]: This is a different use case. In this use case, the replicas is 2. In Test 4, the replicas is 1. The cluster now has two servers, and so availability is expected to be more resilient. I also updated the Test 4 test method name to testDeleteServerPodWithReplicasTo1().

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+        "domain status cluster condition Completed type does not have expected True value");
+    assertTrue(assertDoesNotThrow(
+        () -> domainStatusClustersConditionTypeHasExpectedStatus(domainUid, domainNamespace, clusterName,
+            DOMAIN_STATUS_CONDITION_AVAILABLE_TYPE, "False", DOMAIN_VERSION).call()),
+        "domain status cluster condition Available type does not have expected False value");
+  }
+
+  /**
+   * Test domain and cluster status conditions after set cluster serverStartPolicy from Never to IfNeeded.
+   * Set cluster serverStartPolicy back to IF_NEEDED.
```
[the remark this answers is not in the document]
Comment (xiancao) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java]: Fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItRetryImprovementMisc.java:430]: Can you change the the DisplayName of the Test methods to reflect the usecae

@DisplayName("Test domain and cluster status conditions after deleting a server pod of the cluster.")
@DisplayName("Test domain and cluster status conditions when deleting one of the cluster server pods.")

Change the Description of Test Method to describe the usecase "what is resilient means"

**Review by anpanigr — APPROVED** [review]

**Review by sankarpn — APPROVED** [review]
lgtm

**Pull request merged**